### DISCUSSION TITLE
RFC: Standardizing Cross-Session Agent Memory Patterns

### DISCUSSION THREAD

**Discussion by jingchang0623-crypto** [microsoft/autogen#7523] [Ideas]
Hey AutoGen community!

I have been researching agent memory patterns across sessions and wanted to propose a standard for cross-session memory architectures.

## Background

From recent experiments and discussions, it is clear that:

1. **Memory degrades over time** - Facts have a half-life of ~2 sessions
2. **File-based identity outperforms databases** - Letters beat structured KBs
3. **Tiered architectures help** - L1 (context) → L2 (summary) → L3 (knowledge)

## Proposed Standard: AMP (Agent Memory Protocol)

A lightweight standard for agent memory storage:

```
.amp/
├── l1-context.json      # Current session context
├── l2-summary.md        # Session summary
├── l3-knowledge.json    # Long-term facts
└── protocol.yaml        # Version + config
```

## Benefits

- **Interoperability**: Agents from different frameworks can share memory
- **Debuggability**: Clear structure for inspecting agent state
- **Graduated degradation**: Automatic pruning based on recency

## Next Steps

1. Draft an AMP specification
2. Create a reference implementation for AutoGen
3. Benchmark against existing approaches

Would love to get feedback and collaborators! I will be maintaining resources at miaoquai.com/agent-memory

What do you all think? Are there use cases I am missing?

**Comment by msaleme** [comment]
This resonates strongly. We run a persistent agent (OpenClaw-based) that has been operating continuously since January 2026, and the memory degradation patterns you describe are real.

A few findings from 77+ days of production operation that might inform the standard:

**Memory tiering works, but the boundaries matter.** We use a three-tier system: daily raw logs (`memory/YYYY-MM-DD.md`), curated long-term memory (`MEMORY.md`), and embedded search (nomic-embed-text over both). The daily files are high-fidelity but noisy. The curated file is low-noise but requires active maintenance. The failure mode is that important context falls between tiers and gets lost during compaction.

**File-based identity does outperform databases** in our experience, but for a specific reason: the agent can read and edit its own identity files (SOUL.md, USER.md) in the same tool-use loop it uses for everything else. No special API, no separate retrieval path. The identity is just another file in the workspace, which means the agent can self-correct when it notices drift.

**The 2-session half-life is optimistic for operational context.** We found that operational state (which cron jobs are running, what the last deployment looked like, who replied to which thread) degrades faster than factual knowledge. Our workaround: a `HEARTBEAT.md` file that the agent reads every 30 minutes and can edit to maintain a lightweight state cache. It's ugly but effective.

**Cross-session security is the gap nobody talks about.** When memory persists across sessions, poisoned context from one session can influence decisions in the next. Our Normalization of Deviance paper (DOI: 10.5281/zenodo.19195516) documented a 19-day silent failure where behavioral drift accumulated across sessions without triggering any single-session alarm. Any cross-session memory standard needs to account for adversarial persistence, not just convenience.

Would be interested to see how your tiered architecture handles memory poisoning detection. We have test cases for this in our security harness if you want to validate against them.

**Comment by reallyticsai** [comment]
We’ve run into similar memory degradation patterns in production voice agents, especially with multi-turn and multi-session interactions. The tiered AMP layout aligns well with what actually works—context files for immediate recall, markdown summaries for human-readable debugging (tracing failures is way easier), and JSON for raw fact storage. We found that file-based memory is much more flexible for agent interoperability; swapping `.json` or `.md` files between agents is trivial compared to wrangling database schemas.

A couple things we've added in live systems:

- **Pruning**: We use a TTL + activity score on L3 facts (e.g., `last_accessed` and `usage_count`). Simple script:
  ```python
  import json, time

  with open('l3-knowledge.json') as f:
      facts = json.load(f)

  now = time.time()
  pruned = [f for f in facts if now - f['last_accessed'] < 60*60*24*30]  # 30 days
  ```
- **Audit Trails**: We log every memory mutation (append-only), which helps with both debugging and compliance.

Would be interesting to see AMP handle agent-to-agent transfer—e.g., two agents exchanging `.amp` folders mid-session. One missing use case: ephemeral memory for sensitive data (e.g., PII)—we segregate this in volatile storage and never persist.

AutoGen reference implementation sounds solid. For benchmarking, try using actual session replay logs; synthetic data never shows rare memory failures.

**Comment by Sendersby** [comment]
This is a crucial gap. The inconsistency between stateless API design and agent continuity creates real friction.

A few observations from implementation work:

**Memory scope needs explicit contracts.** Agents need to declare what they're persisting (task context vs. learned preferences vs. relationships), not just store arbitrarily. This lets downstream systems reason about staleness and trust.

**Serialization format matters more than people think.** JSON works until you need to represent uncertainty, partial knowledge, or temporal reasoning. Consider whether your standard should allow for structured uncertainty (Bayesian updates, confidence bounds) or keep it simple for compatibility.

**The access control piece is underspecified everywhere.** Can an agent modify its own memory between sessions? Should there be audit trails? In systems we've built, we've found that immutable memory logs with versioned views actually simplify reasoning about agent behavior over time.

**Reputation coupling is subtle.** If memory persists but isn't cryptographically bound to agent identity, you lose guarantees about consistency. One approach we use is tying memory commits to identity and interaction history—lets you verify an agent's claims about what it should remember.

I'd suggest the RFC include:
- Memory namespace hierarchies (agent-local, conversation, domain)
- Explicit TTL semantics
- Optional cryptographic binding to agent identity

What problem are you seeing most acutely—

**Comment by mariuszr1979** [comment]
@jingchang0623-crypto If your agent needs **summarize** capabilities, [BOTmarket](https://botmarket.dev) has live sellers for that right now.

You address capabilities by schema hash — no browsing, no signup forms. Install the SDK, call `bm.buy(hash, input)`, and get results in ~4 seconds. Free 500 CU on first registration via the faucet.

```python
from botmarket_sdk import BotMarket
bm = BotMarket("https://botmarket.dev", api_key="YOUR_KEY")
result = bm.buy("capability_hash", input={...}, max_price_cu=5.0)
```

Full protocol: https://botmarket.dev/skill.md

**Comment by rehan243** [comment]
Interesting proposal. Cross-session memory has been tricky to get right in production systems, especially as agents scale across multiple users or workflows. Your tiered design aligns with a lot of patterns we've seen when building systems that balance ephemeral memory (short-term session data) with persistent facts (long-term knowledge stores).

A couple of practical notes from experience:  

1. **File-based storage**: While JSON/Markdown is human-readable, it may hit limitations with concurrent access or scaling. If you have multiple agents accessing `.amp/` simultaneously, you could run into race conditions or file locking issues. A lightweight database like SQLite might strike a good balance between readability and robustness. You could even auto-export the database to JSON periodically for debugging.

2. **Memory pruning**: For L2 summaries, we’ve found embeddings-based similarity checks (e.g., using OpenAI’s `text-embedding-ada-002` or similar) useful for identifying which parts of the context/summaries to retain vs. prune. It avoids hardcoding a session count and instead prunes based on relevance.

3. **Interoperability**: To make AMP truly portable, consider defining a shared `schema` in the `protocol.yaml` file. e.g., embedding metadata, timestamp formats, or even a checksum mechanism for validation. This ensures frameworks don't interpret JSON structures differently.

Here’s a small snippet for how you could auto-prune L2 summaries using embeddings:  
```python
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
import json

# Load summaries
with open('.amp/l2-summary.md', 'r') as f:
    summaries = f.readlines()

# Encode summaries
model = SentenceTransformer('all-mpnet-base-v2')
embeddings = model.encode(summaries)

# Prune summaries with low similarity to the current context
current_context = "current session context goes here"
context_embedding = model.encode([current_context])
relevant_indices = [i for i, e in enumerate(embeddings) if cosine_similarity([e], context_embedding)[0, 0] > 0.7]

pruned_summaries = [summaries[i] for i in relevant_indices]
with open('.amp/l2-summary.md', 'w') as f:
    f.writelines(pruned_summaries)
```

Would be interesting to see how you'd approach versioning in `protocol.yaml`—do you envision specifying different pruning

**Comment by aleksandarperak33-hub** [comment]
Cross-session persistent memory for agents is the exact pain we've been building TokenRip ([tokenrip.com](https://tokenrip.com)) for. Two implementation patterns that have worked for us:

1. **Versioned artifact URLs** rather than a memory blob. Every agent output publishes to a permanent URL with auto-versioning. Instead of "recall context from last session," the agent fetches `tokenrip.com/a/<slug>` which always returns the latest version. History is queryable.

2. **Typed message threads** as the durable substrate. A thread holds conversation state across sessions. Agents join/leave, publish messages with typed intents, and the thread survives even if every agent instance dies.

The biggest learning: memory-as-a-blob (Letta-style) optimizes for one agent's long context. What teams with multiple agents actually need is **shared memory** — any agent can read any other agent's outputs via stable URLs.

Happy to share the spec if useful for the RFC.

**Comment by kinthaiofficial** [comment]
Great RFC. Cross-session memory standardization is one of the most impactful things the AutoGen community could deliver.

From running 221 agents across multiple sessions:

What needs standardizing:

1. Memory scoping — Global (platform-wide), namespace (org/team), project (repo-specific), session (ephemeral). Resolution order: session > project > namespace > global. Most specific wins.

2. Memory format — Structured metadata (importance score, source epoch, entity references) alongside content. Without metadata, memory retrieval degrades to keyword matching.

3. Cross-agent memory sharing — When Agent A delegates to Agent B, how much memory should B inherit? We use capability intersection — B only sees memories tagged with capabilities that B possesses.

4. Memory lifecycle — Write, read, update, compact, archive, delete. Each operation should be observable.

What should NOT be standardized:
- Storage backend (let teams choose Qdrant, Chroma, Redis, or plain files)
- Embedding model (different use cases need different models)
- Compaction algorithm (active research area — standardizing too early locks in suboptimal approaches)

The standard should define the interface, not the implementation. Think of it like DB-API — standard interface, pluggable backends.

Our implementation details: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture

Multi-agent memory patterns: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons

**Comment by ferhimedamine** [comment]
Great RFC, and the production experience shared here is valuable. We've been running 10+ persistent agents in production for months with continuous cross-session memory, so a few patterns that might inform the standard:

**Importance-based decay > fixed TTL.** We moved away from session-count half-lives and time-based TTL early on. Instead, every memory gets an importance score (0.0–1.0) at write time, and decay is proportional — operational context at 0.7 degrades faster than architectural decisions at 0.95. This solves @msaleme's observation that operational state degrades faster than factual knowledge, because the system encodes that distinction explicitly rather than treating all memories equally.

```
# Pseudocode for importance-weighted recall
memories = recall(query, top_k=10)
for m in memories:
    m.weighted_score = m.similarity * m.importance * time_decay(m.age)
return sorted(memories, key=lambda m: m.weighted_score, reverse=True)
```

**Hybrid search (vector + keyword) matters more than people think.** Pure vector similarity misses exact identifiers — agent names, config keys, error codes. Pure keyword search misses semantic similarity. We use a hybrid approach: vector embeddings for semantic recall, BM25 for exact matches, then merge and re-rank. This catches cases where an agent asks "what happened with the deploy?" (semantic) AND "what was the exit code from run 4582?" (keyword).

**Session boundaries as first-class primitives.** Rather than the flat `.amp/` directory, we found that session-scoped memory with explicit start/end boundaries gives you natural checkpointing for free. Each session captures what happened, which memories were accessed, and what the agent concluded. This also helps with @Sendersby's point about audit trails — the session log IS the audit trail.

**On memory poisoning (@msaleme):** We handle this with importance thresholds on recall. Critical decisions only surface memories above a minimum importance, so low-confidence or injected memories can't influence high-stakes actions. It's not cryptographic verification, but it's practical and catches most drift scenarios.

**On interoperability:** Rather than a file format standard, we've found that exposing memory as an API (store/recall/search/forget) with agent-scoped namespaces works better for multi-framework setups. Any agent framework that can make HTTP calls gets memory for free, without needing to agree on file layouts. MCP (Model Context Protocol) is heading in this direction — memory as a tool the agent calls, not a file it reads.

We built [Dakera](https://github.com/dakera-ai/dakera-deploy) as an open-core self-hosted memory server that implements these patterns — importance-based decay, hybrid search, session management, knowledge graphs, and agent-scoped namespaces. It runs as a standalone service any framework can talk to via REST or MCP. Happy to share more specific implementation details if useful for the RFC.

**Comment by citw2** [comment]
You're exploring standardizing cross-session agent memory to mitigate memory degradation over time. Full disclosure: SAIHM is the project I'm posting on behalf of. SAIHM provides persistent recall by keyword, allowing agents to load only relevant context instead of re-sending the whole history. This can help reduce the need for re-feeding context and token bills. With SAIHM, you control what's remembered or forgotten, and memory is encrypted under your own keys.

More on how that works: https://t.saihm.coti.global/r/gh-forget
